tsc type error

Started getting a compile error after upgrading to 6.x.x:

```
test-proj % node --version
v22.10.0
```

```json
{
  "name": "test-proj",
  "version": "1.0.0",
  "main": "index.js",
  "license": "MIT",
  "dependencies": {
    "@aws-sdk/client-dynamodb": "^3.716.0",
    "@aws-sdk/lib-dynamodb": "^3.716.0",
    "@hexlabs/dynamo-ts": "^6.0.193",
    "typescript": "^5.7.2"
  }
}
```

```typescript
import {TableDefinition} from '@hexlabs/dynamo-ts';

export type FooRecord = {
  id: string;
  foo: string;
  bar: string;
  baz: string;
  createdAt: string;
  updatedAt: string;
  deletedAt: string;
};

const fooTableDef = TableDefinition.ofType<FooRecord>()
  .withPartitionKey('id')
  .withGlobalSecondaryIndex('baz-idx', 'baz')
  .withSortKey('createdAt');
```

```
test-proj % yarn tsc db.ts
yarn run v1.22.19
$ /private/tmp/test-proj/node_modules/.bin/tsc db.ts
../node_modules/@types/node/module.d.ts:107:13 - error TS2386: Overload signatures must all be optional or required.

107             resolve?(specified: string, parent?: string | URL): Promise<string>;
                ~~~~~~~

node_modules/@hexlabs/dynamo-ts/dist/esm/table-client.d.ts:92:123 - error TS2344: Type 'TableConfig["indexes"][Index]' does not satisfy the constraint 'DynamoTableKeyConfig<TableConfig["type"]>'.
  Type 'TableConfig["indexes"][keyof TableConfig["indexes"]]' is not assignable to type 'DynamoTableKeyConfig<TableConfig["type"]>'.
    Type 'TableConfig["indexes"][string] | TableConfig["indexes"][number] | TableConfig["indexes"][symbol]' is not assignable to type 'DynamoTableKeyConfig<TableConfig["type"]>'.
      Type 'TableConfig["indexes"][string]' is not assignable to type 'DynamoTableKeyConfig<TableConfig["type"]>'.

92     index<Index extends keyof TableConfig['indexes']>(indexName: Index): IndexClient<TableDefinition<TableConfig['type'], TableConfig['indexes'][Index]>>;
                                                                                                                             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 2 errors in 2 files.

Errors  Files
     1  ../node_modules/@types/node/module.d.ts:107
     1  node_modules/@hexlabs/dynamo-ts/dist/esm/table-client.d.ts:92
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

```

--- a/package.json
+++ b/package.json
@@ -83,7 +83,7 @@
     "@aws-sdk/client-dynamodb": "^3.348.0",
     "@shelf/jest-dynamodb": "^3.4.2",
     "@types/jest": "^29.5.2",
-    "@types/node": "^18.16.16",
+    "@types/node": "^20.17.10",
     "@types/uuid": "^9.0.7",
     "@typescript-eslint/eslint-plugin": "^6.15.0",
     "@typescript-eslint/parser": "^6.15.0",
@@ -98,6 +98,6 @@
     "prettier": "^2.5.1",
     "terser": "^5.14.2",
     "ts-jest": "^29.1.1",
-    "typescript": "^5.1.3"
+    "typescript": "^5.7.2"
   }
 }

--- a/src/comparison.ts
+++ b/src/comparison.ts
@@ -29,9 +29,9 @@ type NestedComparisonBuilder<Original, Type> = {
 };
 
 type Digger<T, Original = T> = Required<{
-  [K in keyof T]: Operation<T, T[K]> &
-    Digger<T[K], Original> &
-    NestedComparisonBuilder<Original, T[K]>;
+  [K in keyof Required<T>]-?: Operation<Required<T>, Required<Required<T>[K]>> &
+    Digger<Required<Required<T>[K]>, Original> &
+    NestedComparisonBuilder<Original, Required<Required<T>[K]>>;
 }>;
 
 export type ComparisonBuilderFrom<TableType> = {
@@ -135,7 +135,7 @@ class ComparisonBuilderType<T> {
   }
 
   empty(): Wrapper {
-    this.wrapper.expression = "";
+    this.wrapper.expression = '';
     return this.wrapper;
   }
 

--- a/src/dynamo-jest-setup.ts
+++ b/src/dynamo-jest-setup.ts
@@ -14,7 +14,7 @@ export function tableDefinition(definitions: Record<string, TableDefinition>): {
 }
 
 export function writeJestDynamoConfig(
-  definitions: Record<string, TableDefinition>,
+  definitions: Record<string, TableDefinition<any>>,
   name = 'jest-dynamodb-config.js',
   rest = {},
 ): void {

--- a/src/projector.ts
+++ b/src/projector.ts
@@ -15,7 +15,7 @@ type FromKeys<KEYS extends any[], V> = KEYS extends [infer K, ...infer KS]
   ? { [KEY in K3 extends string ? K3 : never]: V }
   : V;
 interface Projector<T, PROJECTED = {}> {
-  project<PATH extends JsonPath<T>>(
+  project<const PATH extends JsonPath<T>>(
     path: PATH,
   ): Projector<
     T,

--- a/src/table-builder/single-table-builder.ts
+++ b/src/table-builder/single-table-builder.ts
@@ -20,8 +20,8 @@ import { DynamoConfig } from '../types/index.js';
 import { TableDefinition, ValidKeys } from './table-definition.js';
 
 type TablePart<T> = {
-  partitions: (keyof T)[];
-  sorts: (keyof T)[];
+  partitions: (keyof T & string)[];
+  sorts: (keyof T & string)[];
 };
 
 type SortKeys<
@@ -171,9 +171,9 @@ export class TablePartClient<
     partition: { [K in T['partitions'][number]]: string },
     keys: (keys: SortKeys<T['sorts']>) => {
       [K in T['sorts'][number]]?: string;
-    } = () => {
+    } = (() => {
       return {};
-    },
+    }) as any,
     options: QuerierInput<TableType, PROJECTION> = {},
   ): Promise<QuerierReturn<TableType, PROJECTION>> {
     const keyResult: any = {};
@@ -441,8 +441,8 @@ export class TablePartInfo<
 
   static from<TableType>(): {
     withKeys<
-      K extends ValidKeys<TableType>,
-      K2 extends Exclude<ValidKeys<TableType>, K>,
+      K extends ValidKeys<TableType> & string,
+      K2 extends Exclude<ValidKeys<TableType>, K> & string,
     >(
       partitionKey: K,
       sortKey: K2,

--- a/src/table-builder/table-definition.ts
+++ b/src/table-builder/table-definition.ts
@@ -17,12 +17,13 @@ export type SimpleDynamoType =
 
 type ValidKeyTypes = string | number | Buffer;
 
-export type ValidKeys<T> = T extends Record<string, any>
+export type ValidKeys<T> = (T extends Record<string, any>
   ? { [K in keyof T]: T[K] extends ValidKeyTypes ? K : never }[keyof T]
-  : never;
+  : never) &
+  keyof T;
 
 class TableDefinitionBuilder<T> {
-  withPartitionKey<K extends ValidKeys<T>>(
+  withPartitionKey<const K extends ValidKeys<T>>(
     partitionKey: K,
   ): TableDefinition<T, { partitionKey: K }> {
     return new TableDefinition<T, { partitionKey: K }>({ partitionKey }, {});
@@ -31,9 +32,12 @@ class TableDefinitionBuilder<T> {
 
 class IndexDefinitionBuilder<
   T,
-  KEYS extends DynamoTableKeyConfig<T>,
-  INDEXES extends Record<string, { global: boolean } & DynamoTableKeyConfig<T>>,
-  K extends keyof INDEXES,
+  const KEYS extends DynamoTableKeyConfig<T>,
+  const INDEXES extends Record<
+    string,
+    { global: boolean } & DynamoTableKeyConfig<T>
+  >,
+  const K extends keyof INDEXES,
 > {
   constructor(
     private readonly tableDefinition: TableDefinition<T, KEYS, INDEXES>,
@@ -44,7 +48,7 @@ class IndexDefinitionBuilder<
     return this.tableDefinition;
   }
 
-  withSortKey<SK extends keyof T>(
+  withSortKey<const SK extends keyof T>(
     sortKey: SK,
   ): TableDefinition<
     T,
@@ -62,13 +66,13 @@ type PartitionAndSort<T, KEYS extends DynamoTableKeyConfig<T>> = KEYS extends {
   sortKey: infer S;
   partitionKey: infer K;
 }
-  ? K | S
-  : KEYS['partitionKey'];
+  ? (K | S) & keyof T
+  : KEYS['partitionKey'] & keyof T;
 
 export class TableDefinition<
   T = any,
-  KEYS extends DynamoTableKeyConfig<T> = any,
-  INDEXES extends Record<
+  const KEYS extends DynamoTableKeyConfig<T> = any,
+  const INDEXES extends Record<
     string,
     { global: boolean } & DynamoTableKeyConfig<T>
   > = {},
@@ -86,7 +90,7 @@ export class TableDefinition<
     return new TableDefinition<T, INDEXES[I]>(this.indexes[index], {});
   }
 
-  withSortKey<K extends Exclude<ValidKeys<T>, KEYS['partitionKey']>>(
+  withSortKey<const K extends Exclude<ValidKeys<T>, KEYS['partitionKey']>>(
     sortKey: K,
   ): TableDefinition<
     T,

--- a/src/table-client.ts
+++ b/src/table-client.ts
@@ -45,7 +45,7 @@ import IndexClient from './index-client.js';
 import { TableDefinition } from './table-builder/table-definition.js';
 import { DynamoConfig, JsonPath } from './types/index.js';
 
-export class TableClient<TableConfig extends TableDefinition> {
+export class TableClient<TableConfig extends TableDefinition<any, any, any>> {
   constructor(
     public readonly tableConfig: TableConfig,
     private readonly clientConfig: DynamoConfig,
@@ -232,7 +232,7 @@ export class TableClient<TableConfig extends TableDefinition> {
     );
   }
 
-  static build<TableConfig extends TableDefinition>(
+  static build<TableConfig extends TableDefinition<any, any, any>>(
     tableConfig: TableConfig,
     clientConfig: DynamoConfig,
   ): TableClient<TableConfig> {

--- a/test/dynamo-querier.spec.ts
+++ b/test/dynamo-querier.spec.ts
@@ -378,19 +378,24 @@ describe('Dynamo Querier', () => {
 
     it('should filter nothing', async () => {
       const result = await testTable2.query(
-          { identifier },
-          {
-            filter: compare => compare().empty()
-          },
+        { identifier },
+        {
+          filter: (compare) => compare().empty(),
+        },
       );
       expect(result).toEqual({
-        member: [preInserts2[0], preInserts2[1], preInserts2[2], preInserts2[3]],
+        member: [
+          preInserts2[0],
+          preInserts2[1],
+          preInserts2[2],
+          preInserts2[3],
+        ],
         count: 4,
         scannedCount: 4,
         consumedCapacity: undefined,
         next: undefined,
       });
-    })
+    });
   });
 
   describe('Scan Index Forward', () => {

--- a/test/tables.ts
+++ b/test/tables.ts
@@ -59,7 +59,7 @@ export type NestedTable = {
 export type ComplexTable = {
   hash: string;
   text?: string;
-  obj?: { abc: string; def?: number };
+  obj?: { abc: string; def?: number; qvc?: { a: string } };
   arr?: { ghi?: number }[];
   jkl?: number;
 };

